### ISSUE TITLE
Asserts-like mechanism in pyright

### ISSUE THREAD

**Issue by mikez** [microsoft/pyright#8107]
**Is your feature request related to a problem? Please describe.**

It is a common pattern for me to validate input in an external validation function. That keeps the code clean and readable.

In this example, when calling `f()`, I don't expect to have to reraise when something already was validated by `validate_offset`.

```py
def f(offset):
    validate_offset(offset)

    # compute modifier
    number, suffix = int(offset[:-1]), offset[-1]

    if suffix == "d":
        modifier = f"-{number} days"
    elif suffix == "w":
        modifier = f"-{number * 7} days"
    elif suffix == "y":
        modifier = f"-{number} years"
    # else:
    #    raise ValueError() # This is here to make pyright etc. happy; alternative?

    # do something with `modifier` here (omitted)
    modifier


def validate_offset(offset):
    if not isinstance(argument, str):
        raise ValueError(f"Invalid offset: {offset!r}")

    suffix = argument[-1:]  # slicing here to handle empty strings
    if suffix not in ("d", "w", "y"):
        raise ValueError(
            f"Invalid offset: {offset!r}\n"
            f"Please specify a string of the format 'X[d/w/y]' "
            "where X is a non-negative integer followed by 'd', 'w', or 'y' "
            "that indicates days, weeks, or years."
        )
```

**Describe the solution you’d like**

Something akin to the `asserts` mechanism in the Typescript language (see [here](https://stackoverflow.com/a/71625025)) provided through a comment or otherwise, to tell pyright that `modified` will be defined.

**Comment by mikez** [comment]
The folks at pylint told me about `typing.assert_never()`, so maybe that's a pattern that could be used here?

**Comment by mikez** [comment]
That seems to fix it. Closing.
